plugins/redfish/fu-ipmi-device.c

```diff
@@ -0,0 +1,573 @@
+			g_set_error(error, G_IO_ERROR, G_IO_ERROR_FAILED, "poll() error %m");
+			return FALSE;
+		}
+		if (rc == 0) {
+			g_set_error(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "timeout waiting for response "
+				    "(netfn %d, cmd %d)",
+				    netfn,
+				    cmd);
+			return FALSE;
+		}
+
+		if (!(pollfds[0].revents & POLLIN)) {
+			g_set_error_literal(error,
+					    G_IO_ERROR,
+					    G_IO_ERROR_FAILED,
+					    "unexpected status");
+			return FALSE;
+		}
+
+		if (!fu_ipmi_device_recv(self,
+					 &resp_netfn,
+					 &resp_cmd,
+					 &seq,
+					 resp_buf2,
+					 resp_buf2sz,
+					 &resp_len2,
+					 error))
+			return FALSE;
+
+		if (seq != self->seq - 1) {
+			g_debug("out-of-sequence reply: "
+				"expected %ld, got %ld",
+				self->seq,
+				seq);
+			if (g_timer_elapsed(timer, NULL) * 1000.f >= timeout_ms) {
+				g_set_error_literal(error,
+						    G_IO_ERROR,
+						    G_IO_ERROR_FAILED,
+						    "timed out");
+				return FALSE;
+			}
+		} else {
+			if (resp_buf2[0] != IPMI_CC_NO_ERROR) {
+				g_set_error(error,
+					    G_IO_ERROR,
+					    G_IO_ERROR_FAILED,
+					    "CC error: %s [%02x]",
+					    fu_ipmi_device_errcode_to_string(resp_buf2[0]),
+					    resp_buf2[0]);
+				return FALSE;
+			}
+			if (resp_buf != NULL) {
+				if (!fu_memcpy_safe(resp_buf,
+						    resp_bufsz,
+						    0x0, /* dst */
+						    resp_buf2,
+						    resp_buf2sz,
+						    0x01, /* src */
+						    resp_bufsz,
+						    error))
+					return FALSE;
+			}
+			if (resp_len != NULL)
+				*resp_len = resp_len2 - 1;
+			if (g_getenv("FWUPD_REDFISH_VERBOSE") != NULL) {
+				g_debug("IPMI netfn: %02x->%02x, cmd: %02x->%02x",
+					netfn,
+					resp_netfn,
+					cmd,
+					resp_cmd);
+			}
+			break;
+		}
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_ipmi_device_setup(FuDevice *device, GError **error)
+{
+	FuIpmiDevice *self = FU_IPMI_DEVICE(device);
+	gsize resp_len = 0;
+	guint8 resp[16] = {0};
+
+	/* get IPMI versions */
+	if (!fu_ipmi_device_transaction(self,
+					IPMI_NETFN_APP_REQUEST,
+					IPMI_GET_DEVICE_ID_CMD,
+					NULL,
+					0,
+					resp,
+					sizeof(resp),
+					&resp_len,
+					FU_IPMI_DEVICE_TIMEOUT,
+					error))
+		return FALSE;
+	if (resp_len == 11 || resp_len == 15) {
+		guint8 bcd;
+		g_autoptr(GString) str = g_string_new(NULL);
+
+		self->device_id = resp[0];
+		self->device_rev = resp[1];
+		bcd = resp[3] & 0x0f;
+		bcd += 10 * (resp[4] >> 3);
+		/* rev1.rev2.aux_revision */
+		g_string_append_printf(str, "%u.%02u", resp[2], bcd);
+		if (resp_len == 15) {
+			g_string_append_printf(str,
+					       ".%02x%02x%02x%02x",
+					       resp[11],
+					       resp[12],
+					       resp[13],
+					       resp[14]);
+		}
+		fu_device_set_version(device, str->str);
+		bcd = resp[4] & 0x0f;
+		bcd += 10 * (resp[4] >> 4);
+		self->version_ipmi = bcd;
+	} else {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_SUPPORTED,
+			    "failed to retrieve Device ID from IPMI");
```

Comment:
including resp_length in the error perhaps?

Reply:
Good idea, fixed.